src/main/java/org/mskcc/oncokb/curation/util/CdxUtils.java

```diff
@@ -0,0 +1,227 @@
+                rowSpanCount--;
+                continue;
+            }
+            try {
+                rowSpanCount = Integer.parseInt(td.first().attributes().get("rowspan")) - 1;
+            } catch (NumberFormatException e) {}
+            ArrayList<String> tableCells = td.stream().limit(3).map(e -> e.text()).collect(Collectors.toCollection(ArrayList::new));
+
+            // Create CompanionDiagnosticDevice entity
+            CompanionDiagnosticDevice cdx = new CompanionDiagnosticDevice();
+            cdx.setName(tableCells.get(0));
+            Set<String> submissionCodes = Arrays
+                .asList(tableCells.get(1).split("\\s"))
+                .stream()
+                .map(code -> code.trim())
+                .filter(code -> code.length() > 0)
+                .map(code -> code.matches("DEN[0-9]{6}\\/K[0-9]{6}") ? code.split("/")[0] : code) // Special case for FerriScan where number is DEN130012/K124065
+                .collect(Collectors.toSet());
+            cdx.setFdaSubmissions(getFDASubmissionFromHTML(submissionCodes));
+            cdx.setManufacturer(tableCells.get(2));
+            companionDiagnosticDevices.add(cdx);
+        }
+        return companionDiagnosticDevices;
+    }
+
+    // Extract the relevant information from the PMA / 510(k) / HDE pages
+    public Set<FdaSubmission> getFDASubmissionFromHTML(Set<String> fdaSubmissionCodes) {
+        // Check if the fda submission codes are valid and purify input
+        Set<String> purifiedSubmissionCodes = new HashSet<>();
+        for (String code : fdaSubmissionCodes) {
+            code = code.toUpperCase();
+            if (code.matches("([A-Z]+[0-9]+)(\\/(S[0-9]+(-(S[0-9]+))?))?")) {
+                String[] splitCode = code.split("/");
+                String primaryPma = splitCode[0];
+                purifiedSubmissionCodes.add(primaryPma);
+                if (splitCode.length > 1) {
+                    String supplement = splitCode[1];
+                    String[] supplementRange = supplement.split("-");
+                    // Sometimes the PMAs are given as ranges (ie. P990081/S001-S028).
+                    Integer start = Integer.valueOf(supplementRange[0].substring(1));
+                    Integer end = supplementRange.length > 1 ? Integer.valueOf(supplementRange[1].substring(1)) : start;
+                    IntStream
+                        .rangeClosed(start, end)
+                        .forEach(num -> {
+                            String pmaString = String.format("%sS%03d", primaryPma, num);
+                            purifiedSubmissionCodes.add(pmaString);
+                        });
+                }
+            }
+        }
+
+        // Fetch the PMA/510(k)/HDE information from webpage
+        Set<FdaSubmission> fdaSubmissions = new HashSet<>();
+        for (String code : purifiedSubmissionCodes) {
+            String prefix = code.split("[0-9]").length > 0 ? code.split("[0-9]")[0] : null;
```

Comment:
Other option is to use regex to get match group so it can be used directly without splitting twice.

Reply:
I modified your regex slightly to `([A-Z]+[0-9]+)(\/((S[0-9]+)?(-(S[0-9]+))?))?` so that if we have a range (ie P050040/S001-S003), then groups 4 and 6 are S001 and S003, respectively.